### PR TITLE
PRMT-4524

### PR DESCRIPTION
**Pull request by martin-nhs**
- DyanmoDB Integration
- RDS Removal

### CODE REVIEW

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/exceptions/TransferRecordNotPresentException.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package uk.nhs.prm.repo.ehrtransferservice.exceptions;
+
+import uk.nhs.prm.repo.ehrtransferservice.exceptions.base.DatabaseException;
+
+import java.util.UUID;
+
+public class TransferRecordNotPresentException extends DatabaseException {
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/exceptions/TransferRecordNotPresentException.java:7]:
```suggestion
public class ConversationRecordNotPresentException extends DatabaseException {
```

Comment (martin-nhs) [reply src/main/java/uk/nhs/prm/repo/ehrtransferservice/exceptions/TransferRecordNotPresentException.java:7]: And equivalent for `CoreRecordNotPresentException`?

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/exceptions/TransferUnableToUpdateException.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package uk.nhs.prm.repo.ehrtransferservice.exceptions;
+
+import uk.nhs.prm.repo.ehrtransferservice.exceptions.base.DatabaseException;
+
+import java.util.UUID;
+
+public class TransferUnableToUpdateException extends DatabaseException {
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/exceptions/TransferUnableToUpdateException.java:7]:
```suggestion
public class UnableToUpdateConversationException extends DatabaseException {
```

Maybe be more explicit here.

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/exceptions/TransferUnableToUpdateException.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    private static final String EXCEPTION_MESSAGE =
+        "Unable to update transfer record for Inbound Conversation ID %s.";
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/exceptions/TransferUnableToUpdateException.java:9]:
```suggestion
        "Unable to update transfer record for Inbound Conversation ID %s";
```

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/database/TransferRepository.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        if(isInboundConversationPresent(inboundConversationId)) {
+            throw new ConversationAlreadyPresentException(inboundConversationId);
+        }
+
+        tableItem.put(INBOUND_CONVERSATION_ID.name, AttributeValue.builder()
+            .s(event.getConversationId())
+            .build());
+
+        tableItem.put(LAYER.name, AttributeValue.builder()
+            .s(CONVERSATION.name()).build());
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/database/TransferRepository.java:48]:
```suggestion
            .s(CONVERSATION.name())
            .build());
```

File: src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java [hunk 5]
```diff
@@ -121,36 +118,52 @@ void shouldPublishCopcMessageToLargeMessageFragmentTopic() throws IOException {
...
-        var inboundQueueFromMhs = new SimpleAmqpQueue(inboundQueue);
-        inboundQueueFromMhs.sendMessage(smallEhrMessageBody, correlationId);
-
-        var smallEhrObservabilityQueueUrl = sqs.getQueueUrl(smallEhrObservabilityQueueName).getQueueUrl();
+        // given
+        final RepoIncomingEvent repoIncomingEvent = createDefaultRepoIncomingEvent(EHR_CORE_INBOUND_CONVERSATION_ID);
+        final String ehrCoreMessageBody = dataLoader.getDataAsString("RCMR_IN030000UK06");
+        final SimpleAmqpQueue inboundQueueFromMhs = new SimpleAmqpQueue(inboundQueue);
+        final String smallEhrObservabilityQueueUrl = sqs.getQueueUrl(smallEhrObservabilityQueueName).getQueueUrl();
+        final String correlationId = UUID.randomUUID().toString();
```
Comment (MohammadIqbalAD-NHS) [inline src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java:152]: We should try to move away from the old team's terminology of calling a conversation id a correlation id.
```suggestion
        final String conversationId = UUID.randomUUID().toString();
```

File: src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java [hunk 6]
```diff
@@ -121,36 +118,52 @@ void shouldPublishCopcMessageToLargeMessageFragmentTopic() throws IOException {
...
             var receivedMessageHolder = checkMessageInRelatedQueue(smallEhrObservabilityQueueUrl);
-            Assertions.assertTrue(receivedMessageHolder.get(0).getBody().contains(smallEhrMessageBody));
+            Assertions.assertTrue(receivedMessageHolder.get(0).getBody().contains(ehrCoreMessageBody));
             Assertions.assertTrue(receivedMessageHolder.get(0).getMessageAttributes().containsKey("traceId"));
             Assertions.assertTrue(receivedMessageHolder.get(0).getMessageAttributes().containsKey("conversationId"));
         });
     }
 
     @Test
     void shouldPassCorrelationIdToBeSetAsTraceId() throws IOException {
```
Comment (MohammadIqbalAD-NHS) [inline src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java:146]: We should try to move away from the old team's terminology of calling a conversation id a correlation id.
```suggestion
    void shouldPassConversationIdToBeSetAsTraceId() throws IOException {
```

File: src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java [hunk 7]
```diff
@@ -121,36 +118,52 @@ void shouldPublishCopcMessageToLargeMessageFragmentTopic() throws IOException {
...
+
+        inboundQueueFromMhs.sendMessage(ehrCoreMessageBody, correlationId);
 
         await().atMost(10, TimeUnit.SECONDS).untilAsserted(() -> {
             var receivedMessageHolder = checkMessageInRelatedQueue(smallEhrObservabilityQueueUrl);
             var message = receivedMessageHolder.get(0);
-            Assertions.assertTrue(message.getBody().contains(smallEhrMessageBody));
+            Assertions.assertTrue(message.getBody().contains(ehrCoreMessageBody));
             Assertions.assertTrue(message.getMessageAttributes().containsKey("traceId"));
             Assertions.assertEquals(message.getMessageAttributes().get("traceId").getStringValue(), correlationId);
```
Comment (MohammadIqbalAD-NHS) [inline src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java:168]:
```suggestion
            Assertions.assertEquals(message.getMessageAttributes().get("traceId").getStringValue(), conversationId);
```

File: src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java (hunk 5 above)
Comment (martin-nhs) [reply src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java:152]: Let's keep this in mind! 😄

File: src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java (hunk 6 above)
Comment (martin-nhs) [reply src/integration/java/uk/nhs/prm/repo/ehrtransferservice/parsers/ParserBrokerIntegrationTest.java:146]: Let's keep this in mind! 😄

**Pull request merged**